File: DesignPatterns/Structural/Composite/Bloco.java
import java.util.ArrayList;

public class Bloco implements BlocoQuestao {
    private ArrayList<BlocoQuestao> todoparte = new ArrayList<>();
    private String name;

    Bloco(String name){
        this.name = name;
    }

    @Override
    public String getName() {
        return this.name;
    }

    public void add(BlocoQuestao todoparte){
        this.todoparte.add(todoparte);
    }

    public BlocoQuestao getChild(int i){
        return this.todoparte.get(i);
    }

    public void showChildren(){
        System.out.println(this.getName());

        for(int i=0; i<todoparte.size(); i++){
            if(todoparte.get(i) instanceof Questao){
                System.out.println(todoparte.get(i).getName());
            }else{
                ((Bloco) todoparte.get(i)).showChildren();
            } 
        }
    }
    
}

File: Questionario/src/Bloco.ts
import { BlocoQuestao } from "./BlocoQuestao";
import { Questao } from "./Questao";

export class Bloco implements BlocoQuestao {
    private todoparte: BlocoQuestao[] = [];
    private name: string;

    constructor(name: string) {
        this.name = name;
    }

    getName(): string {
        return this.name;
    }

    add(todoparte: BlocoQuestao): void {
        this.todoparte.push(todoparte);
    }

    getChild(i: number): BlocoQuestao {
        return this.todoparte[i];
    }

    showChildren(): void {
        console.log(this.getName());

        for (let i = 0; i < this.todoparte.length; i++) {
            if (this.todoparte[i] instanceof Questao) {
                console.log(this.todoparte[i].getName());
            } else {
                (this.todoparte[i] as Bloco).showChildren();
            }
        }
    }
}
